exporter/exporterhelper/queued_retry_inmemory.go

```diff
@@ -71,49 +71,128 @@ func (qCfg *QueueSettings) Validate() error {
 	return nil
 }
 
+func (qCfg *QueueSettings) getStorageExtension(extensions map[config.ComponentID]component.Extension) (storage.Extension, error) {
+	if qCfg.StorageID != nil {
+		if ext, found := extensions[*qCfg.StorageID]; found {
+			if storageExt, ok := ext.(storage.Extension); ok {
+				return storageExt, nil
+			}
+			return nil, errWrongExtensionType
+		}
+	}
+
+	return nil, nil
+}
+
+func (qCfg *QueueSettings) toStorageClient(ctx context.Context, host component.Host, ownerID config.ComponentID, signal config.DataType) (storage.Client, error) {
+	extension, err := qCfg.getStorageExtension(host.GetExtensions())
+	if err != nil {
+		return nil, err
+	}
+	if extension == nil {
+		return nil, errNoStorageClient
+	}
+
+	client, err := extension.GetClient(ctx, component.KindExporter, ownerID, string(signal))
+	if err != nil {
+		return nil, err
+	}
+
+	return client, err
+}
+
+var (
+	errNoStorageClient    = errors.New("no storage client extension found")
+	errWrongExtensionType = errors.New("requested extension is not a storage extension")
+)
+
 type queuedRetrySender struct {
-	fullName        string
-	cfg             QueueSettings
-	consumerSender  requestSender
-	queue           internal.ProducerConsumerQueue
-	retryStopCh     chan struct{}
-	traceAttributes []attribute.KeyValue
-	logger          *zap.Logger
+	fullName           string
+	id                 config.ComponentID
+	signal             config.DataType
+	cfg                QueueSettings
+	consumerSender     requestSender
+	queue              internal.ProducerConsumerQueue
+	queueStartFunc     internal.PersistentQueueStartFunc
+	retryStopCh        chan struct{}
+	traceAttributes    []attribute.KeyValue
+	logger             *zap.Logger
+	requeuingEnabled   bool
+	requestUnmarshaler internal.RequestUnmarshaler
```

Comment:
Ideally this PR should do one thing: enable the feature. Instead I see modifications of the logic. Can we split the PR into 2 parts:
1. Change whatever logic needs to change and whatever refactoring is necessary. Keep the persistent queue feature experimental.
2. Make persistent queue feature part of stable build.
Otherwise I have a hard time understanding what changes are needed for enabling the feature, which changes are refactorings which don't change the behavior and what changes change the behavior.

Reply:
Yeah, this became too complicated along the way. I'm going to split it into three PRs:
1. Make the queue logic consistent between the in-memory queue and the persistent one. This way enabling the persistent queue is just a change to the build.
2. Enable persistent queue in the build by default.
3. Configuration changes for setting the storage for the persistent queue.

Reply:
The first PR: https://github.com/open-telemetry/opentelemetry-collector/pull/5764. Would you prefer to close this one right now or wait until all the changes are merged individually? @tigrannajaryan @bogdandrutu ?